### PR TITLE
add incremental to ~~jdbc~~ jooq source (and postgres)

### PR DESCRIPTION
**Pull request by cgardens**
## What
* Adds support for incremental syncs to the ~~jdbc~~ jooq-compatible sources. The goal of this PR is to set the base for all jdbc source and to be able to ship the postgres source using incremental.

## How
* Add `JdbcStateManager` to track original and new cursor states as we progress. This should be reusable for all database style source I think.
* Add an iterator (`StateDecoratingIterator`) that handles tracking the max cursor value and emitting the state message at the end of syncing each stream. Note: 1 state message per stream, even if no records are replicated for the stream.
* Decided to take a shortcut here and just do jooq databases for this PR. 
    * Down side of this is that i just copy and pasted the existing `AbstractJdbcSource` for now. 

## Decisions to make
* Is separating the JdbcSource and JooqSource necessary? The hard thing™ here is constructing `WHERE` clauses with the correct syntax to do the comparison we need for the cursor based on the type of the column.  Jooq seems to make that easier (though as you go through the PR, you'll see it is not perfect). What I plan to do next is see what a _just_ jdbc version of this looks like. If it is good enough, I will just get rid of all jooq from the database sources and they can all use one. That's the dream.
    * The problem is made _much_ worse because our data types aren't granular enough. e.g. we treat DateTime as a string. But when we do the query in the db, we need to know that that string is actually a datetime. So the internals of the source can't piggy back off the airbyte data model. This breaks the previously `JdbcSourceAbstraction` because it trusted that our types were good enough.

## Checklist
- [x] **fix generating json2pojo models**
- [x] **add unit tests for state manager**
- [x] **run it in integration tests**
- [x] **add it officially to the pg source and make sure standard tests pass**
- [x] **manual test of PG since we don't have super comprehensive standard tests for incremental yet. (manual testing unfortunately caught a pretty terrible bug. i think this is signal that we need to up our game on test our standard source test)**
- [x] **document how one is supposed to add a catalog for standard source tests.**
- [ ] **bump version**
## Reading Order
* `AbstractJdbcSource`
* `JdbcStateManager`
* `StateDecoratingIterator`
* the rest

### CODE REVIEW

File: airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AbstractJdbcState1.java
Comment (michel-tricot) [inline airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AbstractJdbcState1.java:10]: is the `1` suffix expected?

Comment (cgardens) [reply airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AbstractJdbcState1.java]: hahaha. definitely a wip. it will go away. there was a name collision i was hacking around while developing.

File: airbyte-commons/src/main/java/io/airbyte/commons/stream/MoreStreams.java
Comment (cgardens) [inline airbyte-commons/src/main/java/io/airbyte/commons/stream/MoreStreams.java:35]: will find replace other instances of this in a separate pr.

File: airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/IncrementalUtils.java
Comment (cgardens) [inline airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/IncrementalUtils.java:80]: opinions?

File: airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/JdbcStateManager.java
Comment (cgardens) [inline airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/JdbcStateManager.java:57]: this is a little goofy as opposed to just returning the cursor info. i didn't want to have to worry about defensive copy issues though.

File: airbyte-integrations/connectors/source-jdbc-models/build.gradle
Comment (cgardens) [inline airbyte-integrations/connectors/source-jdbc-models/build.gradle:1]: I was running into a weird issue while trying to nest this model inside of `airbyte-integrations/connectors/source-jdbc` so I pulled it out to be a top level connector while I was developing. I will fix the issue and nest everything that is currently in `airbyte-integrations/connectors/source-jdbc-models` in `airbyte-integrations/connectors/source-jdbc/models`

File: airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AbstractJdbcSource.java
Comment (cgardens) [inline airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AbstractJdbcSource.java:226]: I need to rethink this comparison here. I think for anything but number types this syntax won't work. need to figure how to cast to the database type and then do the comparison. same when calculating `maxCursor` below.

File: airbyte-integrations/bases/standard-source-test/src/main/java/io/airbyte/integrations/standardtest/source/TestSource.java
Comment (cgardens) [inline airbyte-integrations/bases/standard-source-test/src/main/java/io/airbyte/integrations/standardtest/source/TestSource.java:317]: as the todo describes, this is really bad. i don't think we need to block postgres on fixing this, but as part of our goal to "launch" incremental, we need to figure out how to include it into our standard tests more ergonomically and transparently. will add an issue for next week. https://github.com/airbytehq/airbyte/issues/1217

File: airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AbstractJooqSource.java
Comment (cgardens) [inline airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AbstractJooqSource.java:74]: see main description. intentionally alot of copy and paste going on right now. goal next week is to reconsolidate.

File: airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AbstractJooqSource.java
Comment (cgardens) [inline airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AbstractJooqSource.java:334]: again something i want to try to resolve next week.

File: airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/StateDecoratingIterator.java
Comment (cgardens) [inline airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/StateDecoratingIterator.java:70]: There's no guarantee here that our comparators will do the exact same thing as the databases. This is pretty scary, but haven't come up with a bullet proof way of fixing it yet. 

I was thinking about what query I could submit to the database to do this for me. Doing it as individual queries would be too expensive, but if I can batch them somehow it might be okay. Haven't decided if this is a right now problem or not. Open to input.

File: airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AbstractJooqSource.java
Comment (sherifnada) [inline airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AbstractJooqSource.java:218]: rename to stateStream?

File: airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AbstractJooqSource.java
Comment (sherifnada) [inline airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AbstractJooqSource.java:191]: the logic right now seems to be: if there was ever a cursor, do incremental. else do fullrefresh. 

Shouldn't it be the other way around i.e: if there is a current cursor AND that cursor is the same as the initial cursor do incremental otherwise FR? might be missing something.

File: airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/IncrementalUtils.java
Comment (sherifnada) [inline airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/IncrementalUtils.java:65]: can this comparison be done DB side?

**Review by sherifnada — COMMENTED** [review]
Main comments are around: 

1. might be missing something, but I'm unsure if the current dynamic of current vs. initial cursor is correct
1. why do we have two methods of discovery? can we have just one to avoid risk of them going out of sync? 
1. is it possible/needed to paginate full refresh queries to the db? ideally incremental too. 
1. is it possible to do cursor comparison DB side? 

Rest of the comments are style or nits

File: airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AbstractJooqSource.java
Comment (cgardens) [reply airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AbstractJooqSource.java:74]: yes.

File: airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AbstractJooqSource.java
Comment (cgardens) [reply airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AbstractJooqSource.java:230]: i don't think that this should be an issue. postgres should do all of this in a streaming fashion so it shouldn't technically cause problems.

File: airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AbstractJooqSource.java
Comment (cgardens) [reply airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AbstractJooqSource.java:119]: i think it does. if you follow discover => getTables => discoverInternal. so they use the same code path. discover just has to wrap it in our data model, while internally we don't wrap it so we can use the more granularly defined types.

Comment (sherifnada) [reply airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AbstractJooqSource.java:119]: ok so I'm missing a detail then ;)

File: airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AbstractJooqSource.java
Comment (cgardens) [reply airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AbstractJooqSource.java:191]: i think you're right. i'll double check this.

File: airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/IncrementalUtils.java
Comment (cgardens) [reply airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/IncrementalUtils.java:65]: if we can we should. see: https://github.com/airbytehq/airbyte/pull/1172/files#r536510322. i have some performance concerns around it.

File: airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AbstractJooqSource.java
Comment (cgardens) [reply airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AbstractJooqSource.java:244]: i'll double check.

File: airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/StateDecoratingIterator.java
Comment (cgardens) [reply airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/StateDecoratingIterator.java:77]: np

File: airbyte-integrations/connectors/source-jdbc/src/test/java/io/airbyte/integrations/source/jdbc/JooqSource.java
Comment (cgardens) [reply airbyte-integrations/connectors/source-jdbc/src/test/java/io/airbyte/integrations/source/jdbc/JooqSource.java:39]: sure. i will clarify.

File: airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AbstractJooqSource.java
Comment (cgardens) [reply airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AbstractJooqSource.java:246]: nope.

File: airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AbstractJooqSource.java
Comment (cgardens) [reply airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AbstractJooqSource.java:220]: that's not accurate because this variable gets reused by the full refresh branch.

File: airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AbstractJooqSource.java
Comment (cgardens) [reply airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/AbstractJooqSource.java:246]: you're right. i'm using initial cursor field here when i should just be using the cursor field. i will fix this and add a test to catch it.

File: airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/IncrementalUtils.java
Comment (cgardens) [reply airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/IncrementalUtils.java:65]: 😏  `select MAX(col1) from ( values ('a'), ('b'), ('c')) as x (col1);`

Comment (cgardens) [reply airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/IncrementalUtils.java:65]: 😏😏😏😏😏🤡 🥨 😏   `ctx.select(DSL.max(DSL.field("col1"))).from(DSL.values(DSL.row("1"), DSL.row("3"), DSL.row("2")).as("t", "col1"))`

Comment (cgardens) [reply airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/IncrementalUtils.java:65]: will decide tomorrow if i'm going to do this in this iteration. but it's nifty.

File: airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/StateDecoratingIterator.java
Comment (michel-tricot) [inline airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/StateDecoratingIterator.java:91]: NIT: remove new line

**Review by michel-tricot — COMMENTED** [review]
It is not ready to merge right?

**Comment by cgardens** [comment]
correct still iterating on this based on sherif's feedback.

File: airbyte-integrations/bases/standard-source-test/src/main/java/io/airbyte/integrations/standardtest/source/TestSource.java
Comment (cgardens) [reply airbyte-integrations/bases/standard-source-test/src/main/java/io/airbyte/integrations/standardtest/source/TestSource.java:313]: we can. would prefer to be explicit since we have two enums with the same name floating around. it's hard for me to remember which is which if one is `EnumName.BLAH` and the other is `BLAH`

**Comment by cgardens** [comment]
This is ready for another pass. I am working on the last todos mentioned above, but I don't think any of them will make or break your review.

File: airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/IncrementalUtils.java
Comment (cgardens) [reply airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/IncrementalUtils.java:63]: going to hold off on this. i think it is low risk for right now and i don't want to put the work in here if it looks like i'm going to replace it with a jdbcsource anyway.

File: airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/JdbcStateManager.java
Comment (cgardens) [reply airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/JdbcStateManager.java:83]: That's not what is happening here. None of these are failure states. Given that it is a valid configuration, these are all possible states that need to be handled. Let me know if you you see a case in here that should be a failure case. I've reviewed each of them and they all seem reasonable to me.

**Review by sherifnada — APPROVED** [review]

**Pull request merged**
